zebrad/src/commands/seed.rs

```diff
@@ -0,0 +1,185 @@
+//! `seed` subcommand - runs a dns seeder
+
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+
+use abscissa_core::{config, Command, FrameworkError, Options, Runnable};
+use futures::channel::oneshot;
+use tower::{buffer::Buffer, Service, ServiceExt};
+use zebra_network::{AddressBook, BoxedStdError, Request, Response};
+
+use crate::{config::ZebradConfig, prelude::*};
+
+/// Whether our `SeedService` is poll_ready or not.
+#[derive(Debug)]
+enum SeederState {
+    /// Waiting for the address book to be shared with us via the oneshot channel.
+    AwaitingAddressBook(oneshot::Receiver<Arc<Mutex<AddressBook>>>),
+    /// Address book received, ready to service requests.
+    Ready(Arc<Mutex<AddressBook>>),
+}
+
+#[derive(Debug)]
+struct SeedService {
+    state: SeederState,
+}
+
+impl Service<Request> for SeedService {
+    type Response = Response;
+    type Error = BoxedStdError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        debug!("SeedService.state: {:?}", self.state);
+
+        match self.state {
+            SeederState::Ready(_) => return Poll::Ready(Ok(())),
+            SeederState::AwaitingAddressBook(ref mut rx) => match rx.try_recv() {
+                Err(e) => {
+                    error!("SeedService oneshot sender dropped: {:?}", e);
+                    return Poll::Ready(Err(e.into()));
+                }
+                Ok(None) => {
+                    debug!("SeedService got a message with `None` in it. 🤔");
+                    return Poll::Pending;
+                }
+                Ok(Some(address_book)) => {
+                    info!(
+                        "SeedService received address_book via oneshot {:?}",
+                        address_book
+                    );
+                    self.state = SeederState::Ready(address_book);
+                    return Poll::Ready(Ok(()));
+                }
+            },
+        }
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        info!("SeedService handling a request: {:?}", req);
+
+        let response = match (req, &self.state) {
```

Comment:
Because the `Service` contract requires callers to call `poll_ready` before `call`, it's OK to unwrap the address book as, e.g.,
```
let address_book = if let SeederState::Ready(address_book) {
    address_book
} else {
    panic!("SeedService::call without SeedService::poll_ready");
}
```
and doing this instead of the tuple match on `(req, &self.state)` ensures that non-`GetPeers` requests (some client sent stuff) are handled differently than violation of the `Service` contract (programming error).